panic while running benchmarks

```
⟩ go version
go version go1.14 darwin/amd64
⟩ git rev-parse HEAD
284ba06683916885c62c6839c87032d9ad1f3fad
⟩ go test -v -count=1 -run=NONE -bench  ./...
goos: darwin
goarch: amd64
pkg: github.com/cockroachdb/pebble
BenchmarkBatchSet
BenchmarkIndexedBatchSet
BenchmarkBatchSetDeferred
BenchmarkIndexedBatchSetDeferred
BenchmarkCommitPipeline
BenchmarkCommitPipeline/parallel=1
BenchmarkCommitPipeline/parallel=1-16  	 1000000	      1016 ns/op	  15.75 MB/s
BenchmarkCommitPipeline/parallel=2
BenchmarkCommitPipeline/parallel=2-16  	 1000000	      1039 ns/op	  15.40 MB/s
BenchmarkCommitPipeline/parallel=4
BenchmarkCommitPipeline/parallel=4-16  	 1000000	      1009 ns/op	  15.85 MB/s
BenchmarkCommitPipeline/parallel=8
BenchmarkCommitPipeline/parallel=8-16  	 1000000	      1111 ns/op	  14.41 MB/s
BenchmarkCommitPipeline/parallel=16
BenchmarkCommitPipeline/parallel=16-16 	 1206015	       976 ns/op	  16.39 MB/s
BenchmarkCommitPipeline/parallel=32
BenchmarkCommitPipeline/parallel=32-16 	 1000000	      1086 ns/op	  14.74 MB/s
BenchmarkCommitPipeline/parallel=64
BenchmarkCommitPipeline/parallel=64-16 	 1000000	      1260 ns/op	  12.70 MB/s
BenchmarkCommitPipeline/parallel=128
BenchmarkCommitPipeline/parallel=128-16         	 1000000	      1264 ns/op	  12.65 MB/s
BenchmarkDelete
BenchmarkDelete/delete
BenchmarkDelete/delete-16                       	      30	  39930403 ns/op
BenchmarkDelete/single-delete
BenchmarkDelete/single-delete-16                	      31	  40606620 ns/op
BenchmarkIteratorSeekGE
BenchmarkIteratorNext
BenchmarkIteratorPrev
BenchmarkLevelIterSeekGE
BenchmarkLevelIterSeekGE/restart=16
BenchmarkLevelIterSeekGE/restart=16/count=5
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x14 pc=0x447609e]

goroutine 19673 [running]:
github.com/cockroachdb/pebble/vfs.(*memFile).Close(0xc00096c0c0, 0x48a85a0, 0xc00096c0c0)
	/Users/raul/W/pl/pebble/vfs/mem_fs.go:608 +0x2e
panic(0x46959e0, 0x4cf7c00)
	/usr/local/Cellar/go/1.14/libexec/src/runtime/panic.go:973 +0x396
github.com/cockroachdb/pebble/vfs.(*memFile).Close(0xc00096c0f0, 0x48a85a0, 0xc00096c0f0)
	/Users/raul/W/pl/pebble/vfs/mem_fs.go:608 +0x2e
panic(0x46959e0, 0x4cf7c00)
	/usr/local/Cellar/go/1.14/libexec/src/runtime/panic.go:973 +0x396
github.com/cockroachdb/pebble/vfs.(*memFile).Close(0xc00096c150, 0x48a85a0, 0xc00096c150)
	/Users/raul/W/pl/pebble/vfs/mem_fs.go:608 +0x2e
panic(0x46959e0, 0x4cf7c00)
	/usr/local/Cellar/go/1.14/libexec/src/runtime/panic.go:973 +0x396
github.com/cockroachdb/pebble/vfs.(*memFile).Close(0xc00096c180, 0x48a85a0, 0xc00096c180)
	/Users/raul/W/pl/pebble/vfs/mem_fs.go:608 +0x2e
panic(0x46959e0, 0x4cf7c00)
	/usr/local/Cellar/go/1.14/libexec/src/runtime/panic.go:973 +0x396
github.com/cockroachdb/pebble/vfs.(*memFile).Close(0xc00096c1b0, 0x48a85a0, 0xc00096c1b0)
	/Users/raul/W/pl/pebble/vfs/mem_fs.go:608 +0x2e
github.com/cockroachdb/pebble.buildLevelIterTables(0xc0002e2700, 0x8000, 0x10, 0x5, 0xc00009f080, 0x5, 0x5, 0xc00096c420, 0x5, 0x5, ...)
	/Users/raul/W/pl/pebble/level_iter_test.go:415 +0xd8b
github.com/cockroachdb/pebble.BenchmarkLevelIterSeekGE.func1.1(0xc0002e2700)
	/Users/raul/W/pl/pebble/level_iter_test.go:427 +0x64
testing.(*B).runN(0xc0002e2700, 0x1)
	/usr/local/Cellar/go/1.14/libexec/src/testing/benchmark.go:191 +0xe8
testing.(*B).run1.func1(0xc0002e2700)
	/usr/local/Cellar/go/1.14/libexec/src/testing/benchmark.go:231 +0x57
created by testing.(*B).run1
	/usr/local/Cellar/go/1.14/libexec/src/testing/benchmark.go:224 +0x7d
exit status 2
FAIL	github.com/cockroachdb/pebble	16.912s
```

--- a/internal/arenaskl/skl_test.go
+++ b/internal/arenaskl/skl_test.go
@@ -811,6 +811,7 @@ func BenchmarkIterPrev(b *testing.B) {
 	}
 
 	it := l.NewIter(nil, nil)
+	_, _ = it.Last()
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
 		key, _ := it.Prev()

--- a/internal/base/iterator.go
+++ b/internal/base/iterator.go
@@ -29,7 +29,8 @@ import "fmt"
 //
 // The relative positioning methods can be used in conjunction with any of the
 // absolute positioning methods with one exception: SeekPrefixGE does not
-// support reverse iteration via Prev.
+// support reverse iteration via Prev. It is undefined to call relative
+// positioning methods without ever calling an absolute positioning method.
 //
 // InternalIterators can optionally implement a prefix iteration mode. This
 // mode is entered by calling SeekPrefixGE and exited by any other absolute

--- a/level_iter_test.go
+++ b/level_iter_test.go
@@ -350,7 +350,7 @@ func TestLevelIterSeek(t *testing.T) {
 
 func buildLevelIterTables(
 	b *testing.B, blockSize, restartInterval, count int,
-) ([]*sstable.Reader, manifest.LevelSlice, [][]byte) {
+) ([]*sstable.Reader, manifest.LevelSlice, [][]byte, func()) {
 	mem := vfs.NewMem()
 	files := make([]vfs.File, count)
 	for i := range files {
@@ -358,7 +358,6 @@ func buildLevelIterTables(
 		if err != nil {
 			b.Fatal(err)
 		}
-		defer f.Close()
 		files[i] = f
 	}
 
@@ -388,7 +387,6 @@ func buildLevelIterTables(
 
 	opts := sstable.ReaderOptions{Cache: NewCache(128 << 20)}
 	defer opts.Cache.Unref()
-
 	readers := make([]*sstable.Reader, len(files))
 	for i := range files {
 		f, err := mem.Open(fmt.Sprintf("bench%d", i))
@@ -398,6 +396,12 @@ func buildLevelIterTables(
 		readers[i], err = sstable.NewReader(f, opts)
 		if err != nil {
 			b.Fatal(err)
+		}
+	}
+
+	cleanup := func() {
+		for _, r := range readers {
+			require.NoError(b, r.Close())
 		}
 	}
 
@@ -412,7 +416,7 @@ func buildLevelIterTables(
 		key, _ = iter.Last()
 		meta[i].Largest = *key
 	}
-	return readers, manifest.NewLevelSlice(meta), keys
+	return readers, manifest.NewLevelSlice(meta), keys, cleanup
 }
 
 func BenchmarkLevelIterSeekGE(b *testing.B) {
@@ -424,7 +428,8 @@ func BenchmarkLevelIterSeekGE(b *testing.B) {
 				for _, count := range []int{5} {
 					b.Run(fmt.Sprintf("count=%d", count),
 						func(b *testing.B) {
-							readers, metas, keys := buildLevelIterTables(b, blockSize, restartInterval, count)
+							readers, metas, keys, cleanup := buildLevelIterTables(b, blockSize, restartInterval, count)
+							defer cleanup()
 							newIters := func(
 								file manifest.LevelFile, _ *IterOptions, _ *uint64,
 							) (internalIterator, internalIterator, error) {
@@ -454,7 +459,8 @@ func BenchmarkLevelIterNext(b *testing.B) {
 				for _, count := range []int{5} {
 					b.Run(fmt.Sprintf("count=%d", count),
 						func(b *testing.B) {
-							readers, metas, _ := buildLevelIterTables(b, blockSize, restartInterval, count)
+							readers, metas, _, cleanup := buildLevelIterTables(b, blockSize, restartInterval, count)
+							defer cleanup()
 							newIters := func(
 								file manifest.LevelFile, _ *IterOptions, _ *uint64,
 							) (internalIterator, internalIterator, error) {
@@ -487,7 +493,8 @@ func BenchmarkLevelIterPrev(b *testing.B) {
 				for _, count := range []int{5} {
 					b.Run(fmt.Sprintf("count=%d", count),
 						func(b *testing.B) {
-							readers, metas, _ := buildLevelIterTables(b, blockSize, restartInterval, count)
+							readers, metas, _, cleanup := buildLevelIterTables(b, blockSize, restartInterval, count)
+							defer cleanup()
 							newIters := func(
 								file manifest.LevelFile, _ *IterOptions, _ *uint64,
 							) (internalIterator, internalIterator, error) {

--- a/mem_table_test.go
+++ b/mem_table_test.go
@@ -398,7 +398,7 @@ func BenchmarkMemTableIterSeekGE(b *testing.B) {
 func BenchmarkMemTableIterNext(b *testing.B) {
 	m, _ := buildMemTable(b)
 	iter := m.newIter(nil)
-
+	_, _ = iter.First()
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
 		key, _ := iter.Next()
@@ -412,7 +412,7 @@ func BenchmarkMemTableIterNext(b *testing.B) {
 func BenchmarkMemTableIterPrev(b *testing.B) {
 	m, _ := buildMemTable(b)
 	iter := m.newIter(nil)
-
+	_, _ = iter.Last()
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
 		key, _ := iter.Prev()

--- a/merging_iter_test.go
+++ b/merging_iter_test.go
@@ -261,7 +261,7 @@ func TestMergingIterCornerCases(t *testing.T) {
 
 func buildMergingIterTables(
 	b *testing.B, blockSize, restartInterval, count int,
-) ([]*sstable.Reader, [][]byte) {
+) ([]*sstable.Reader, [][]byte, func()) {
 	mem := vfs.NewMem()
 	files := make([]vfs.File, count)
 	for i := range files {
@@ -269,7 +269,6 @@ func buildMergingIterTables(
 		if err != nil {
 			b.Fatal(err)
 		}
-		defer f.Close()
 		files[i] = f
 	}
 
@@ -322,7 +321,11 @@ func buildMergingIterTables(
 			b.Fatal(err)
 		}
 	}
-	return readers, keys
+	return readers, keys, func() {
+		for _, r := range readers {
+			require.NoError(b, r.Close())
+		}
+	}
 }
 
 func BenchmarkMergingIterSeekGE(b *testing.B) {
@@ -334,7 +337,8 @@ func BenchmarkMergingIterSeekGE(b *testing.B) {
 				for _, count := range []int{1, 2, 3, 4, 5} {
 					b.Run(fmt.Sprintf("count=%d", count),
 						func(b *testing.B) {
-							readers, keys := buildMergingIterTables(b, blockSize, restartInterval, count)
+							readers, keys, cleanup := buildMergingIterTables(b, blockSize, restartInterval, count)
+							defer cleanup()
 							iters := make([]internalIterator, len(readers))
 							for i := range readers {
 								var err error
@@ -363,7 +367,8 @@ func BenchmarkMergingIterNext(b *testing.B) {
 				for _, count := range []int{1, 2, 3, 4, 5} {
 					b.Run(fmt.Sprintf("count=%d", count),
 						func(b *testing.B) {
-							readers, _ := buildMergingIterTables(b, blockSize, restartInterval, count)
+							readers, _, cleanup := buildMergingIterTables(b, blockSize, restartInterval, count)
+							defer cleanup()
 							iters := make([]internalIterator, len(readers))
 							for i := range readers {
 								var err error
@@ -395,7 +400,8 @@ func BenchmarkMergingIterPrev(b *testing.B) {
 				for _, count := range []int{1, 2, 3, 4, 5} {
 					b.Run(fmt.Sprintf("count=%d", count),
 						func(b *testing.B) {
-							readers, _ := buildMergingIterTables(b, blockSize, restartInterval, count)
+							readers, _, cleanup := buildMergingIterTables(b, blockSize, restartInterval, count)
+							defer cleanup()
 							iters := make([]internalIterator, len(readers))
 							for i := range readers {
 								var err error

